extensions/auth/portability-auth-harness-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/harness/AuthTestDriver.java

```diff
@@ -0,0 +1,89 @@
+package org.dataportabilityproject.auth.microsoft.harness;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+import org.dataportabilityproject.auth.microsoft.MicrosoftAuthDataGenerator;
+import org.dataportabilityproject.spi.gateway.types.AuthFlowConfiguration;
+import org.dataportabilityproject.types.transfer.auth.TokenAuthData;
+
+import java.awt.*;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Objects;
+
+import static java.lang.System.getProperty;
+
+/**
+ *
+ */
+public class AuthTestDriver {
+    private String clientId;
+    private String secret;
+
+    private String callbackHost;
+    private String callbackBase;
+    private String authRetrievalUrl;
+
+    public AuthTestDriver() {
+        this.callbackHost = Objects.requireNonNull(getProperty("callbackHost"), "Callback host");
+        String callbackPort = Objects.requireNonNull(getProperty("callbackPort"), "Callback port");
+        this.clientId = Objects.requireNonNull(getProperty("clientId"), "Client ID");
+        this.secret = Objects.requireNonNull(getProperty("secret"), "Client secret");
+
+        callbackBase = "https://" + callbackHost + ":" + callbackPort;
+        authRetrievalUrl = callbackBase + "/code";
+    }
+
+    public static void main(String... args) throws Exception {
```

Comment:
I agree. That's just for me to "manually" test until I can come up with proper auth stubs (testing OAuth is a pain when your internet connection is reliant on a cheap modem that doesn't support port forwarding :(). 

I can remove it now, wait until later to remove it, or document that it will be removed. Let me know what you think works best.

Reply:
Gotcha, fine with me to either leave a TODO to move/remove it, or move it to a test directory

Reply:
If you want to make this an integration test, looks like the recommended approach in gradle is to make a separate "source dir" for integration tests rather than lumping it in with unit tests https://guides.gradle.org/migrating-from-maven/?_ga=2.198959343.1349228063.1517424826-1039443295.1516912197#integration_tests (just came across this as I was reading about gradle)

Reply:
Thanks for the pointer. I like the ideas of separate phases and dependencies for integration tests.

My original plan was just to remove this method since it is really only designed as a temporary way to test against a live MS instance to get the code in a state where it works.

Moving forward, I'd like to get proper integration tests in place for this extension:

1. A set of fast running unit tests that can be executed as part of any build without opening network connections (if possible, there will be some exceptions for localhost) or other resources. I should be able to stub out the OKHttp client to do this and have a connection that is driven by a local mock. I was going to take a similar approach with the Cassandra driver for the store (which already uses the SCassandra stub).

2. At some point we should do different levels of end-to-end integration testing. One would be testing individual extensions against test accounts, which the Gradle approach in the link will come in handy for. Another type of integration test would be to launch a complete system (workers and storage) and drive transfer. The easiest way to do this would probably be to build a test runner or Gradle plugin that bootstraps the data transfer system with provided config. If we created a Gradle plugin, other people could use it for internal testing as well.

Reply:
We can add these suggestions to the existing test issue or have multiple ones for each kind of test.